### PR TITLE
[FIXED JENKINS-30080] Fix the deadlock between scheduleBuild() and NodeProvisioner updates

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Jenkins JIRA is down, so I have not referenced the issue yet. Please find the description below.

<code>NodeProvisioner#getPendingLaunches()</code> and <code>NodeProvisioner#update()</code> may get deadlocked if somebody retrieves a list of pending launches in the build scheduler. It 
also applies to all other methods guarded by <code>provisioningLock</code>, because <code>NodeProvisioner#update()</code> locks the queue after locking <code>provisioningLock</code>.

The idea of the fix is to reverse the locking order in <code>NodeProvisioner#update()</code>. It requires spinning off the Strategy apply() handlers to another locked section.  I also had to propagate and handle exceptions.

Stacktrace example:

```
SCM polling for hudson.model.FreeStyleProject@2913c875[FED-Trunk-Javascript-Stubserver] locked on java.util.concurrent.locks.ReentrantLock$NonfairSync@51ab5668 (owned by Computer.threadPoolForRemoting [#3]):
     at sun.misc.Unsafe.park(Native Method)
     at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:834)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:867)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1197)
     at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:214)
     at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:290)
     at hudson.slaves.NodeProvisioner.getPendingLaunches(NodeProvisioner.java:158)
     at XXXX.CloudImpl$ProvisionStrategy.shouldSchedule(CloudImpl.java:181)
     at hudson.model.Queue.schedule2(Queue.java:588)
     at jenkins.model.ParameterizedJobMixIn.scheduleBuild2(ParameterizedJobMixIn.java:115)
     at jenkins.model.ParameterizedJobMixIn.scheduleBuild2(ParameterizedJobMixIn.java:104)
     at hudson.model.AbstractProject.scheduleBuild2(AbstractProject.java:870)
     at hudson.model.AbstractProject.scheduleBuild2(AbstractProject.java:852)
     at hudson.model.Project.scheduleBuild2(Project.java:113)
     at hudson.triggers.SCMTrigger$Runner.run(SCMTrigger.java:552)
     at hudson.util.SequentialExecutionQueue$QueueEntry.run(SequentialExecutionQueue.java:118)
     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
     at java.util.concurrent.FutureTask.run(FutureTask.java:262)
     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
     at java.lang.Thread.run(Thread.java:745)
, Computer.threadPoolForRemoting [#3] locked on java.util.concurrent.locks.ReentrantLock$NonfairSync@2a0b6368 (owned by SCM polling for hudson.model.FreeStyleProject@2913c875[FED-Trunk-Javascript-Stubserver]):
     at sun.misc.Unsafe.park(Native Method)
     at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:834)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:867)
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1197)
     at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:214)
     at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:290)
     at hudson.model.Queue._withLock(Queue.java:1205)
     at hudson.model.Queue.withLock(Queue.java:1143)
     at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:208)
     at hudson.slaves.NodeProvisioner.access$000(NodeProvisioner.java:57)
     at hudson.slaves.NodeProvisioner$1.run(NodeProvisioner.java:177)
     at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
     at java.util.concurrent.FutureTask.run(FutureTask.java:262)
     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
     at java.lang.Thread.run(Thread.java:745)
```

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Also CCed @stephenc , who is an author of the original code

**Comment by stephenc** [comment]
:bug: this is not the way to fix this IMHO. There was a specific reason for the locking order. With this new locking order you can get a race where the Queue lock operations take a long time to complete and multiple threads stack up and then fall through the strategy being over-applied as they all run through fighting for the strategy lock with the same effective state, resulting in double or worse provisioning of the excess load.

The correct solution, to my mind, is CAS on an immutable state object so that getPendingLaunches is lock-free

**Comment by oleg-nenashev** [comment]
@stephenc 
In such case we may need to modify all 3 public methods, because they may be called from queue maintenance methods as well. After that the <code>provisioningLock</code> will become useless at all.

**Comment by oleg-nenashev** [comment]
Probably we could add a Strategy change queue to avoid the excess load.

File: core/src/main/java/hudson/slaves/NodeProvisioner.java [hunk 1]
```diff
@@ -776,10 +803,26 @@ public long getRecurrencePeriod() {
...
+                        "Jenkins has not been started, or was already shut down");
+                return;
+            }
+            try {
+                h.unlabeledNodeProvisioner.update();
+            } catch (Exception ex) {
+                LOGGER.log(Level.SEVERE, "Cannot update the unlabeled node provisioner ", ex);
+            }
+                
+            for( Label l : h.getLabels() ) {
```
Comment (KostyaSha) [inline core/src/main/java/hudson/slaves/NodeProvisioner.java]: ..

**Comment by stephenc** [comment]
In my opinion #1785 is a better fix of this issue

**Comment by oleg-nenashev** [comment]
Created https://issues.jenkins-ci.org/browse/JENKINS-30080

**Pull request closed without merging**